nixos/tests/password-option-override-ordering.nix

```diff
@@ -0,0 +1,151 @@
+            hashedPassword = hashed_sha512crypt;
+            password = password1;
+            hashedPasswordFile = (pkgs.writeText "hashed_bcrypt" hashed_bcrypt).outPath; # Expect override of everything above
+          };
+
+          # Show hashedPassword -> password -> hashedPasswordFile -> initialPassword is false
+          # to explicitly show the following lib.trace warning in users-groups.nix is wrong:
+          # ```
+          # The user 'root' has multiple of the options
+          # `hashedPassword`, `password`, `hashedPasswordFile`, `initialPassword`
+          # & `initialHashedPassword` set to a non-null value.
+          # The options silently discard others by the order of precedence
+          # given above which can lead to surprising results. To resolve this warning,
+          # set at most one of the options above to a non-`null` value.
+          # ```
+          users.cat = {
+            isNormalUser = true;
+            hashedPassword = hashed_sha512crypt;
+            password = password1;
+            hashedPasswordFile = (pkgs.writeText "hashed_bcrypt" hashed_bcrypt).outPath;
+            initialPassword = password2; # lib.trace message implies this overrides everything above
+          };
+
+          # Show hashedPassword -> password -> hashedPasswordFile -> initialHashedPassword is false
+          # to explicitly show the lib.trace shown above is wrong
+          users.dan = {
+            isNormalUser = true;
+            hashedPassword = hashed_sha512crypt;
+            initialPassword = password2;
+            password = password1;
+            hashedPasswordFile = (pkgs.writeText "hashed_bcrypt" hashed_bcrypt).outPath;
+            initialHashedPassword = hashed_yeshash; # lib.trace message implies this overrides everything above
+          };
+        };
+      };
+
+    testScript = ''
+      with subtest("alice user has correct password"):
+        print(machine.succeed("getent passwd alice"))
+        assert "${hashed_sha512crypt}" not in machine.succeed("getent shadow alice"), "alice user password is not correct"
+
+      with subtest("bob user has correct password"):
+        print(machine.succeed("getent passwd bob"))
+        assert "${hashed_bcrypt}" in machine.succeed("getent shadow bob"), "bob user password is not correct"
+
+      with subtest("cat user has correct password"):
+        print(machine.succeed("getent passwd cat"))
+        assert "${hashed_bcrypt}" in machine.succeed("getent shadow cat"), "cat user password is not correct"
+
+      with subtest("dan user has correct password"):
+        print(machine.succeed("getent passwd dan"))
+        assert "${hashed_bcrypt}" in machine.succeed("getent shadow dan"), "dan user password is not correct"
+
+      with subtest("greg user has correct password"):
+        print(machine.succeed("getent passwd greg"))
+        assert "${hashed_sha512crypt}" not in machine.succeed("getent shadow greg"), "greg user password is not correct"
```

Comment:
We could match the output of the hashed password from `getent` against the password set for greg here I think.

Also, it's a little weird that you return `passwd` rather than shadow above?

Reply:
> Can only check exact match when providing the hash yourself.

Sure, but there are tools that you can use to verify the hash against the plaintext password.

> By return you mean what is printed ya? I switched them all to print getent shadow. Makes sense.

Correct.

Reply:
> > Can only check exact match when providing the hash yourself.
> 
> Sure, but there are tools that you can use to verify the hash against the plaintext password.

Ok. I'll add that